This fixes and issue with Duo bypass resulting in an authentication error

atticus88 opened the pull request:
This change fixes and issue that occurs if a user is set to bypass on the Duo Administrator Console.
It would result in an "INVALID_AUTHENTICATION_CONTEXT" failure.

```
 <cas:authenticationFailure code="INVALID_AUTHENTICATION_CONTEXT">
```

I was seeing the error in  CAS 5.2.4 and above, I did not test lower versions.

This is a result of the DefaultDuoMultifactorAuthenticationProvider returning that it is not supported.
I have change it to return true if the account is set to bypass.

I have tested this change with my setup and and has fixed the issue.

mmoayyed commented:
Than you very much for the patch.

I don't think this is the right fix. Can you describe the circumstances the lead to `INVALID_AUTHENTICATION_CONTEXT`? I believe changes are more likely required in the place where the authn context is validated, aka https://github.com/apereo/cas/pull/3482

atticus88 commented:
```
CAS Version: 5.2.4
CAS Commit Id: 67d7e128955437619534b5af4819f2379b934353
CAS Build Date/Time: 2018-04-13T21:46:16Z
Spring Boot Version: 1.5.12.RELEASE
------------------------------------------------------------
Java Home: /usr/lib/jvm/java-8-openjdk-amd64/jre
Java Vendor: Oracle Corporation
Java Version: 1.8.0_181
JVM Free Memory: 95 MB
JVM Maximum Memory: 990 MB
JVM Total Memory: 213 MB
JCE Installed: Yes
------------------------------------------------------------
OS Architecture: amd64
OS Name: Linux
OS Version: 4.4.0-130-generic
OS Date/Time: 2018-08-14T09:22:13.889
OS Temp Directory: /tmp/tomcat8-tomcat8-tmp
------------------------------------------------------------
```

Here are the logs of a login of a user that is in bypass on the Duo side:
```
2018-08-14 09:23:09,480 INFO [PolicyBasedAuthenticationManager] - <Authenticated principal [klintduotest] with attributes [{displayName=[klintduotest], GOBTPAC_EXTERNAL_USER=[klintduotest], memberOf=[CN=auth_duoOU=groups,DC=example,DC=com], msDS-UserPasswordExpiryTimeComputed=[132094270306397245], pwdLastSet=[131778046306397245], sAMAccountName=[klintduotest]}] via credentials [[klintduotest]].>
2018-08-14 09:23:09,481 INFO [Slf4jLoggingAuditTrailManager] - <Audit trail record BEGIN
=============================================================
WHO: klintduotest
WHAT: Supplied credentials: [klintduotest]
ACTION: AUTHENTICATION_SUCCESS
APPLICATION: CAS
WHEN: Tue Aug 14 09:23:09 MDT 2018
CLIENT IP ADDRESS: 192.168.1.176
SERVER IP ADDRESS: 192.168.1.25
=============================================================

>
2018-08-14 09:23:09,786 INFO [Slf4jLoggingAuditTrailManager] - <Audit trail record BEGIN
=============================================================
WHO: klintduotest
WHAT: [event=mfa-duo,timestamp=Tue Aug 14 09:23:09 MDT 2018,source=RegisteredServicePrincipalAttributeMultifactorAuthenticationPolicyEventResolver]
ACTION: AUTHENTICATION_EVENT_TRIGGERED
APPLICATION: CAS
WHEN: Tue Aug 14 09:23:09 MDT 2018
CLIENT IP ADDRESS: 192.168.1.176
SERVER IP ADDRESS: 192.168.1.25
=============================================================

>
2018-08-14 09:23:10,198 INFO [Slf4jLoggingAuditTrailManager] - <Audit trail record BEGIN
=============================================================
WHO: klintduotest
WHAT: TGT-1-*********************************************************H4tIURR-L4-te-casdev1
ACTION: TICKET_GRANTING_TICKET_CREATED
APPLICATION: CAS
WHEN: Tue Aug 14 09:23:10 MDT 2018
CLIENT IP ADDRESS: 192.168.1.176
SERVER IP ADDRESS: 192.168.1.25
=============================================================

>
2018-08-14 09:23:10,255 INFO [DefaultCentralAuthenticationService] - <Granted ticket [ST-1-Fetae-ngqx0SA86AFbVFmdkssFM-te-casdev1] for service [https://hadoopdev1.weber.edu/casdev1/] and principal [klintduotest]>
2018-08-14 09:23:10,256 INFO [Slf4jLoggingAuditTrailManager] - <Audit trail record BEGIN
=============================================================
WHO: klintduotest
WHAT: ST-1-Fetae-ngqx0SA86AFbVFmdkssFM-te-casdev1 for https://hadoopdev1.weber.edu/casdev1/
ACTION: SERVICE_TICKET_CREATED
APPLICATION: CAS
WHEN: Tue Aug 14 09:23:10 MDT 2018
CLIENT IP ADDRESS: 192.168.1.176
SERVER IP ADDRESS: 192.168.1.25
=============================================================

>
2018-08-14 09:23:10,382 INFO [Slf4jLoggingAuditTrailManager] - <Audit trail record BEGIN
=============================================================
WHO: klintduotest
WHAT: ST-1-Fetae-ngqx0SA86AFbVFmdkssFM-te-casdev1
ACTION: SERVICE_TICKET_VALIDATED
APPLICATION: CAS
WHEN: Tue Aug 14 09:23:10 MDT 2018
CLIENT IP ADDRESS: 192.168.1.35
SERVER IP ADDRESS: 192.168.1.25
=============================================================

>
2018-08-14 09:23:10,393 WARN [DefaultAuthenticationContextValidator] - <No satisfied multifactor authentication providers are recorded in the current authentication context.>
```

CAS Response:
```
<cas:serviceResponse xmlns:cas='http://www.yale.edu/tp/cas'>
   <cas:authenticationFailure code="INVALID_AUTHENTICATION_CONTEXT">The validation request for [&#39;ST-1-Fetae-ngqx0SA86AFbVFmdkssFM-te-casdev1&#39;] cannot be satisfied. The request is either unrecognized or unfulfilled.</cas:authenticationFailure>
</cas:serviceResponse>
```


Bypass (Not Working)

```
2018-08-15 09:12:38,663 DEBUG [BaseDuoSecurityAuthenticationService] - <Contacting Duo to inquire about username [klintduotest]>
2018-08-15 09:12:38,921 DEBUG [BaseDuoSecurityAuthenticationService] - <Received Duo admin response [{"response": {"result": "allow", "status_msg": "Logging you in automatically..."}, "stat": "OK"}]>
2018-08-15 09:12:38,921 DEBUG [DefaultDuoMultifactorAuthenticationProvider] - <Found duo user account status [org.apereo.cas.adaptors.duo.DuoUserAccount@49754dc2[status=ALLOW,enrollPortalUrl=<null>]] for [klintduotest]>
2018-08-15 09:12:38,921 DEBUG [DefaultDuoMultifactorAuthenticationProvider] - <Account status is set for allow/bypass for [klintduotest]>
```

Normal User not in bypass by Duo:

```
2018-08-15 09:13:27,126 DEBUG [BaseDuoSecurityAuthenticationService] - <Contacting Duo to inquire about username [klintholmes]>
2018-08-15 09:13:27,222 DEBUG [BaseDuoSecurityAuthenticationService] - <Received Duo admin response [{"response": {"devices": REMOVED}>
2018-08-15 09:13:27,222 DEBUG [DefaultDuoMultifactorAuthenticationProvider] - <Found duo user account status [org.apereo.cas.adaptors.duo.DuoUserAccount@a0af2d9[status=AUTH,enrollPortalUrl=<null>]] for [klintholmes]>
2018-08-15 09:13:27,222 DEBUG [BaseDuoSecurityAuthenticationService] - <Contacting Duo to inquire about username [klintholmes]>
2018-08-15 09:13:27,335 DEBUG [BaseDuoSecurityAuthenticationService] - <Received Duo admin response [{"response": {"devices": REMOVED}>
2018-08-15 09:13:27,335 DEBUG [DefaultDuoMultifactorAuthenticationProvider] - <Found duo user account status [org.apereo.cas.adaptors.duo.DuoUserAccount@5e565608[status=AUTH,enrollPortalUrl=<null>]] for [klintholmes]>
2018-08-15 09:13:27,336 DEBUG [BaseDuoSecurityAuthenticationService] - <Contacting Duo to inquire about username [klintholmes]>
2018-08-15 09:13:27,424 DEBUG [BaseDuoSecurityAuthenticationService] - <Received Duo admin response [{"response": {"devices": REMOVED}>
2018-08-15 09:13:39,281 DEBUG [BasicDuoSecurityAuthenticationService] - <Calling DuoWeb.verifyResponse with signed request token '[AUTH|REMOVED|REMOVED|REMOVED|REMOVED'>
2018-08-15 09:13:39,281 DEBUG [DuoAuthenticationHandler] - <Response from Duo verify: [klintholmes]>
2018-08-15 09:13:39,281 INFO [DuoAuthenticationHandler] - <Successful Duo authentication for [klintholmes]>
2018-08-15 09:13:39,281 INFO [PolicyBasedAuthenticationManager] - <Authenticated principal [klintholmes] with attributes [{}] via credentials [[[username=klintholmes,signedDuoResponse=AUTH|REMOVED|REMOVED|REMOVED|REMOVED]]].>
```

Found some more logs that might help find where the issue is happening. Looks like DefaultDuoMultifactorAuthenticationProvider does not support the the bypass on the Duo side.
```
2018-08-15 15:19:13,943 DEBUG [AbstractMultifactorAuthenticationProvider] - <[DefaultDuoMultifactorAuthenticationProvider] voted does not support this authentication request>
.......
2018-08-15 15:19:14,052 DEBUG [DefaultAuthenticationContextValidator] - <Attempting to match requested authentication context [mfa-duo] against [[]]>
2018-08-15 15:19:14,052 DEBUG [DefaultAuthenticationContextValidator] - <No authentication context could be determined based on authentication attribute [authnContextClass]>
2018-08-15 15:19:14,052 WARN [DefaultAuthenticationContextValidator] - <No satisfied multifactor authentication providers are recorded in the current authentication context.>
2018-08-15 15:19:16,572 DEBUG [WebApplicationServiceFactory] - <No service is specified in the request. Skipping service creation>
```

tsschmidt commented:
I think the issue comes form being bypassed by Duo and not a CAS process.  So Auth Provider needs to set bypass attributes in the authn when Duo returns allow from in the determineDuoUserAccount action.

mmoayyed commented:
Thank you very much for the logs, @atticus88. 

The bit about `No satisfied multifactor authentication providers are recorded in the current authentication context.` is likely where the issue is. 

As @tsschmidt says, if bypass is in fact happening, certain attributes are recorded to indicate that as part of the collected authentication, and it's likely the validator is not processing those attributes correctly, not honoring bypass. The fix will likely need to be part of the validator, and perhaps https://github.com/apereo/cas/pull/3482 does it already; I will try to test this particular pull request against 5.3 and 5.2 and see how it might behave and we can take it from there.

atticus88 commented:
Thanks, let me know if you need anything else from me.

tsschmidt commented:
In DefaultDuoMultifactorAuthenticationProvider, I think this will fix the issue:
```
if (acct.getStatus() == DuoUserAccountAuthStatus.ALLOW) {
        LOGGER.debug("Account status is set for allow/bypass for [{}]", principal);
        authentication.addAttribute(AUTHENTICATION_ATTRIBUTE_BYPASS_MFA, Boolean.TRUE);
        authentication.addAttribute(AUTHENTICATION_ATTRIBUTE_BYPASS_MFA_PROVIDER, getId());
        return false;
}
```
Another option would be create a DuoBypassProvider the extends DefaultMFABypassProvider and adds a REST call to Duo to see if the user is marked as bypass.  I can't remember how to log in to our Duo admin so I can set up a test for this

atticus88 commented:
I tested the above option and it does authenticate successfully with the user in bypass.

tsschmidt commented:
I think the entire method supportsInternal(...) can go away in DefaultDuoMultifactorAuthenticationProvider and this can be handled in DetermineDuoUserAccount.  Currently we call this endpoint to Duo twice for each duo provider in the resolved list.  I also have a need for this change, soon, and was getting ready put PR together to refactor DetermineDuoUserAccount for handling Duo outages.  Should have some to review by tomorrow, targeting 5.3.x

mmoayyed commented:
@atticus88 Thank you for the patch and suggestion again. Closing in favor https://github.com/apereo/cas/pull/3493.

Please keep'em coming :)

mmoayyed commented:
A smaller version of the patch https://github.com/apereo/cas/pull/3493 is made available in 5.2.x, and should be ready for testing in the current snapshot.
